src/kmeans.jl

```diff
@@ -269,89 +280,82 @@ function update_centers!(
+                         assignments::Vector{Int}, to_update::Vector{Bool},
+                         centers::Matrix{TC}, cweights::Vector{TW}
+                         ) where {TX<:Real,TC<:AbstractFloat,TW<:Real}
     d, n = size(X)
     k = size(centers, 2)
 
     # initialize center weights
-    cweights[to_update] .= 0.0
+    cweights[to_update] .= zero(TW)
 
     # accumulate columns
-    @inbounds for j = 1:n
+    @inbounds for j ∈ 1:n
+        # skip samples with negative weights or assigned to a center
+        # that doesn't need to be updated
         wj = weights[j]
-        if wj > 0
-            cj = assignments[j]
-            1 <= cj <= k || error("assignment out of boundary.")
-            if to_update[cj]
-                if cweights[cj] > 0
-                    for i = 1:d
-                        centers[i, cj] += X[i, j] * wj
-                    end
-                else
-                    for i = 1:d
-                        centers[i, cj] = X[i, j] * wj
-                    end
+        cj = assignments[j]
+        if to_update[cj] && wj > 0
+            if cweights[cj] > 0
+                for i ∈ 1:d
+                    centers[i, cj] += X[i, j] * wj
+                end
+            else
+                for i ∈ 1:d
+                    centers[i, cj] = X[i, j] * wj
                 end
-                cweights[cj] += wj
             end
+            cweights[cj] += wj
         end
     end
 
     # sum ==> mean
-    @inbounds for j = 1:k
+    @inbounds for j ∈ 1:k
         if to_update[j]
-            for i = 1:d
-                centers[i, j] /= cweights[j]
+            cj = cweights[j]
+            for i ∈ 1:d
+                centers[i, j] /= cj
             end
         end
     end
 end
 
 
-#
-#  Re-picks centers that get no samples assigned to them.
-#
-function repick_unused_centers(
-    X::AbstractMatrix{T},       # in: the sample set (d x n)
-    costs::Vector{T},           # in: the current assignment costs (n)
-    centers::AbstractMatrix{T}, # to be updated: the centers (d x k)
-    unused::Vector{Int}         # in: set of indices of centers to be updated
-    ) where T<:Real
+####
+#### Re-pick centers that get no samples assigned to them.
+####
 
+function repick_unused_centers(X::AbstractMatrix{TX}, costs::Vector{TD},
+                               centers::Matrix{TC}, unused::Vector{Int},
+                               distance::SemiMetric
+                               ) where {TX<:Real,TC<:AbstractFloat,TD<:Real}
     # pick new centers using a scheme like kmeans++
     ds = similar(costs)
     tcosts = copy(costs)
     n = size(X, 2)
 
-    for i in unused
+    for i ∈ unused
         j = wsample(1:n, tcosts)
         tcosts[j] = 0
         v = view(X, :, j)
-        centers[:, i] = v
-
-        colwise!(ds, SqEuclidean(), v, X)
+        # NOTE REFACT: centers is of type TC which is a supertype of TX
+        copyto!(centers[:, i], v)
```

Comment:
This won't work AFAICT since it will modify a copy. What's wrong with the old code?

Reply:
this line was hit in my local tests worked fine and seem to decrease allocations but happy to revert if it's unclear.

Reply:
`copyto!(centers[:, i]) = v`: `centers[:, i]` creates a new array, `copyto!()` modifies it, and then it is goes straight to GC, since nobody references it.
`centers[:, i] = v` is lowered by the compiler into `setindex!(centers, :, i) = v`, which seems right to me; no extra allocation. What exactly was the problem with this line?
(`SqEuclidean` is a nice catch :))

Reply:
ok fair enough will revert, thanks!